Add tests for FormatterServices.GetUninitializedObject

hughbe opened the pull request:
- non runtime types
- array, pointer, ref types
- abstract/interface types
- open generic types
- byref like types
- shared generic instantiations
- COM objects
- nullable types
- default constructor
- static constructor (non beforefieldinit)
- static field (beforefieldinit)
- static constructor throwing

Fixes #16995

jkotas commented on src/System.Runtime.Serialization.Formatters/tests/FormatterServicesTests.cs:
```diff
@@ -29,21 +34,260 @@ public void GetSerializableMembers_Interface()
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, "The full .NET framework supports GetUninitializedObject for shared generic instances")]
+        public void GetUniniitalizedObject_SharedGenericInstance_NetCore_ThrowsNotSupportedException()
+        {
+            Type canonType = Type.GetType("System.__Canon");
```
I do not think it is worth the pain to test the behavior for internal CoreLib types. The behavior for internal CoreLib types will differ from runtime to runtime. These tests would need to have yet another customization for uataop or for Mono.

hughbe replied:
I think this is an important test, as it exercises a direct codepath in the coreclr implementation of GetUninitializedObject. And, it calls out a difference between netfx and netcoreapp, whereby GetSafeUninitializedObject doesn't throw in netfx, but does in netcoreapp due to runtime differences.

What I have done, however, is changed the condition for the skipping to be `~TargetFrameworkMonikers.Netcoreapp` to avoid errors for uataop and mono

jkotas commented on src/System.Runtime.Serialization.Formatters/tests/FormatterServicesTests.cs:
```diff
@@ -29,21 +34,260 @@ public void GetSerializableMembers_Interface()
+
+        [Fact]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, "The full .NET framework supports GetUninitializedObject for shared generic instances")]
+        public void GetUniniitalizedObject_SharedGenericInstance_NetCore_ThrowsNotSupportedException()
```
Typo: GetUniniitalizedObject

hughbe replied:
fixed

jkotas commented on src/System.Runtime.Serialization.Formatters/tests/FormatterServicesTests.cs:
```diff
@@ -29,21 +34,260 @@ public void GetSerializableMembers_Interface()
             Assert.Throws<ArgumentNullException>("type", () => FormatterServices.GetSafeUninitializedObject(null));
         }
 
+#if netcoreapp
```
You can create a non-runtime type by creating your local NonRumtimeType that inherits from Type. It will work everywhere. The netcoreapp ifdef would not be needed.

hughbe replied:
This is great to know! Looks like the default constructor for `Type` was added back recently. I've extracted this to a common helper, as I know I can update `System.Runtime.Tests` to use this for no runtime types everywhere

jkotas commented on src/System.Runtime.Serialization.Formatters/tests/FormatterServicesTests.cs:
```diff
@@ -29,21 +34,260 @@ public void GetSerializableMembers_Interface()
+            yield return new object[] { typeof(RuntimeArgumentHandle) };
+            yield return new object[] { typeof(TypedReference) };
+
+            yield return new object[] { typeof(Span<int>) };
```
The full framework runtime is not aware of Span type special properties today, so this test will fail on full framework.

hughbe replied:
These tests aren't run on the full .NET framework - they are skipped.

morganbr replied:
I don't think uapaot or corert know about Span either

jkotas replied:
> uapaot or corert know about Span either

They will know soon.

jkotas commented on src/System.Runtime.Serialization.Formatters/tests/FormatterServicesTests.cs:
```diff
@@ -29,21 +34,260 @@ public void GetSerializableMembers_Interface()
+            Assert.Throws<NotSupportedException>(() => FormatterServices.GetSafeUninitializedObject(typeof(COMObject)));
+        }
+
+        public class COMObject : ContextBoundObject
```
ContextBoundObjects are not COMObjects.

hughbe replied:
Shows how little I know about the runtime :D.

I've renamed this, and also added  a test for an actual COM object

morganbr commented on src/System.Runtime.Serialization.Formatters/tests/FormatterServicesTests.cs:
```diff
@@ -29,21 +31,263 @@ public void GetSerializableMembers_Interface()
+
+        public static IEnumerable<object[]> GetUninitializedObject_ByRefLikeType_NetCore_TestData()
+        {
+            yield return new object[] { Type.GetType("System.ByReference`1[System.Int32]") };
```
Reflection on ByReference will be blocked on uapaot and I'm not sure what corert will do. @jkotas , do you have any ideas on how to get a byref-like Type on those runtimes?

jkotas replied:
I had the same comment above that it is questionable for corefx tests to depend on internal CoreLib types.

> do you have any ideas on how to get a byref-like Type on those runtimes?

TypedReference or (soon) Span<T> are fine byref-like types.

morganbr commented on src/System.Runtime.Serialization.Formatters/tests/FormatterServicesTests.cs:
```diff
@@ -29,21 +31,263 @@ public void GetSerializableMembers_Interface()
+        [SkipOnTargetFramework(~TargetFrameworkMonikers.NetFramework, "The full .NET framework partially supports GetUninitializedObject for shared generic instances")]
+        public void GetUninitializedObject_SharedGenericInstance_Netfx_InitializesValue()
+        {
+            Type canonType = Type.GetType("System.__Canon");
```
@davidwrighton , does this work on uapaot?

jkotas replied:
Right, it won't work. I had the same comment above that it is questionable for corefx tests to depend on internal CoreLib types.

hughbe commented:
I've addressed your comments about relying on runtime specific code such as Span<T>, ReadOnlySpan<T> and ByReference<T>, by making these coreclr specific. This way, uapaot and Mono won't be broken.

hughbe commented:
@dotnet-bot Test Innerloop Windows_NT Debug x86 Build and Test

danmosemsft commented:
@jkotas can you please take another look so we can merge?

jkotas commented on src/System.Runtime.Serialization.Formatters/tests/FormatterServicesTests.cs:
```diff
@@ -29,21 +31,257 @@ public void GetSerializableMembers_Interface()
+        }
+
+        [Theory]
+        [MemberData(nameof(GetUninitializedObject_ByRefLikeType_TestData))]
```
This does not look right. `ArgIterator`, etc. should get exceptions thrown on full framework.

jkotas replied:
The list can be inline for this particular case.

hughbe replied:
I've ran the netfx tests `msbuild /T:BuildAndTest /P:TargetGroup=netfx` and they pass, so I guess exceptions aren't throw.

Also tested locally:
```cs

class Program
{
    static void Main()
    {
        FormatterServices.GetUninitializedObject(typeof(ArgIterator));
        FormatterServices.GetUninitializedObject(typeof(RuntimeArgumentHandle));
        FormatterServices.GetUninitializedObject(typeof(TypedReference));

        Console.WriteLine("Done");
        Console.ReadLine();
    }
}
```

Prints `Done` - i.e. no exception thrown

jkotas replied:
I see. I would fix the comment to say "full .NET framework has bug that allows creating byref-like types". E.g. if you do anything interesting with TypedReference created this way, you will tend to get crashed during GC.

jkotas replied:
Also, I do not think we need a test in corefx that verifies the buggy netfx behavior (https://github.com/dotnet/corefx/issues/17723#issuecomment-290763235).

hughbe replied:
For the record I disagree greatly - I think netfx specific tests document the difference in behaviour better than any reference to an issue or pr or any comment could.
It's not about moving internal .net tests to corefx, but more about documenting the difference. This is important considering each difference makes it riskier to port-to-core. Death by a thousand cuts or whatnot.
Never mind, I've removed the netfx specific tests from this PR

jkotas commented on src/System.Runtime.Serialization.Formatters/tests/FormatterServicesTests.cs:
```diff
@@ -29,21 +31,257 @@ public void GetSerializableMembers_Interface()
+        {
+            yield return new object[] { typeof(Span<int>) };
+            yield return new object[] { typeof(ReadOnlySpan<int>) };
+            yield return new object[] { Type.GetType("System.ByReference`1[System.Int32]") };
```
I am still not very happy about testing these internal CoreLib types. The tests that do it are high maintenance. They have to be fixed in creative ways again and again. I do not think it is the right tradeoff.

If you would like more interesting cases covered here, you can add this instead of ``ByRefeference<T>`:

```
struct Reader
{
     Span<byte> _bytes;
     int position;
}
```

This is a public normal type, no runtime internals.

hughbe replied:
Funky - I've replaced the `ByReference` use with this example.

hughbe commented:
@jkotas awesome! are you merging or waiting for a second view

joperezr commented:
LGTM, thanks @hughbe !
